test.cc

```diff
@@ -0,0 +1,266 @@
+  EXPECT_NEAR(0.0, result.value, 1e-3);
+  ASSERT_EQ(2u, result.point.size());
+  EXPECT_NEAR(0.0, result.point[0], 1e-3);
+  EXPECT_NEAR(0.0, result.point[1], 1e-3);
+}
+
+TEST(FNCAS, OptimizationOfAPolynomialUsingBacktrackingGD) {
+  const auto result = 
+    fncas::GradientDescentOptimizerBT<PolynomialFunctionToOptimize>().Optimize(std::vector<double>({5.0, 20.0}));
+  EXPECT_NEAR(0.0, result.value, 1e-3);
+  ASSERT_EQ(2u, result.point.size());
+  EXPECT_NEAR(0.0, result.point[0], 1e-3);
+  EXPECT_NEAR(0.0, result.point[1], 1e-3);
+}
+
+TEST(FNCAS, OptimizationOfAPolynomialUsingConjugateGradient) {
+  const auto result = 
+    fncas::ConjugateGradientOptimizer<PolynomialFunctionToOptimize>().Optimize(std::vector<double>({5.0, 20.0}));
+  EXPECT_NEAR(0.0, result.value, 1e-6);
+  ASSERT_EQ(2u, result.point.size());
+  EXPECT_NEAR(0.0, result.point[0], 1e-6);
+  EXPECT_NEAR(0.0, result.point[1], 1e-6);
+}
+
+TEST(FNCAS, OptimizationOfRosenbrockUsingConjugateGradient) {
+  const auto result = 
+    fncas::ConjugateGradientOptimizer<RosenbrockFunctionToOptimize>().Optimize(std::vector<double>({-3.0, -4.0}));
+  EXPECT_NEAR(0.0, result.value, 1e-6);
+  ASSERT_EQ(2u, result.point.size());
+  EXPECT_NEAR(1.0, result.point[0], 1e-6);
+  EXPECT_NEAR(1.0, result.point[1], 1e-6);
+}
+
+TEST(FNCAS, OptimizationOfHimmelbaluUsingCojugateGradient) {
+  fncas::ConjugateGradientOptimizer<HimmelblauFunctionToOptimize> optimizer;
+  const auto min1 = optimizer.Optimize(std::vector<double>({5.0, 5.0}));
+  EXPECT_NEAR(0.0, min1.value, 1e-6);
+  ASSERT_EQ(2u, min1.point.size());
+  EXPECT_NEAR(3.0, min1.point[0], 1e-6);
+  EXPECT_NEAR(2.0, min1.point[1], 1e-6);
+
+  const auto min2 = optimizer.Optimize(std::vector<double>({-3.0, 5.0}));
+  EXPECT_NEAR(0.0, min2.value, 1e-6);
+  ASSERT_EQ(2u, min2.point.size());
+  EXPECT_NEAR(-2.805118, min2.point[0], 1e-6);
+  EXPECT_NEAR(3.131312, min2.point[1], 1e-6);
+
+  const auto min3 = optimizer.Optimize(std::vector<double>({-5.0, -5.0}));
+  EXPECT_NEAR(0.0, min3.value, 1e-6);
+  ASSERT_EQ(2u, min3.point.size());
+  EXPECT_NEAR(-3.779310, min3.point[0], 1e-6);
+  EXPECT_NEAR(-3.283186, min3.point[1], 1e-6);
+
+  const auto min4 = optimizer.Optimize(std::vector<double>({5.0, -5.0}));
+  EXPECT_NEAR(0.0, min4.value, 1e-6);
+  ASSERT_EQ(2u, min4.point.size());
+  EXPECT_NEAR(3.584428, min4.point[0], 1e-6);
+  EXPECT_NEAR(-1.848126, min4.point[1], 1e-6);
+}
+
+TEST(FNCAS, NaiveGDvsBacktrackingGDOnRosenbrockFunction1000Steps) {
+  fncas::OptimizerParameters params;
+  params.SetValue("max_steps", 1000);
+  const auto result_naive = 
+    fncas::GradientDescentOptimizer<RosenbrockFunctionToOptimize>(params).Optimize(std::vector<double>({-3.0, -4.0}));
+  const auto result_bt = 
+    fncas::GradientDescentOptimizerBT<RosenbrockFunctionToOptimize>(params).Optimize(std::vector<double>({-3.0, -4.0}));
+  const double x0_err_n = std::abs(result_naive.point[0] - 1.0);
```

Comment:
If `NaN` may appear here because of subtracting the numbers that are to small, why not reduce the number of iterations significantly and deal with comparable numbers instead?